pkg/controller/instance/instance_controller.go

```diff
@@ -97,63 +100,71 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 				// since its linking to a bad FV.
 				return false
 			}
-			// Identify plan to be executed by this change
+
+			// Identify plan to be executed by this change.
 			var planName string
-			var ok bool
+			var planFound bool
 			if old.Spec.FrameworkVersion != new.Spec.FrameworkVersion {
-				// Its an Upgrade!
-				_, ok = fv.Spec.Plans["upgrade"]
-				if !ok {
-					_, ok = fv.Spec.Plans["update"]
-					if !ok {
-						_, ok = fv.Spec.Plans["deploy"]
-						if !ok {
-							log.Println("InstanceController: Could not find any plan to use for upgrade")
-							return false
-						}
-						ok = true // TODO: Do we need this here?
-						planName = "deploy"
-					} else {
-						planName = "update"
+				// Its an upgrade!
+				names := []string{"upgrade", "update", "deploy"}
+				for _, n := range names {
+					if _, planFound = fv.Spec.Plans[n]; planFound {
+						planName = n
+						break
 					}
-				} else {
-					planName = "upgrade"
+				}
+				if !planFound {
+					log.Printf("InstanceController: Could not find any plan to use to upgrade instance %v", new.Name)
+					return false
 				}
 			} else if !reflect.DeepEqual(old.Spec, new.Spec) {
 				for k := range parameterDifference(old.Spec.Parameters, new.Spec.Parameters) {
-					// Find the right parameter in the FV
+					// Find the spec of the updated parameter.
+					paramFound := false
 					for _, param := range fv.Spec.Parameters {
 						if param.Name == k {
-							planName = param.Trigger
-							ok = true
+							paramFound = true
+
+							if param.Trigger != "" {
+								planName = param.Trigger
+								planFound = true
+							}
+							// TODO: what should we do if more than one parameter was updated?
 						}
 					}
-					if !ok {
-						log.Printf("InstanceController: Instance %v updated parameter %v, but parameter not found in FrameworkVersion %v\n", new.Name, k, fv.Name)
-					} else if planName == "" {
-						_, ok = fv.Spec.Plans["update"]
-						if !ok {
-							_, ok = fv.Spec.Plans["deploy"]
-							if !ok {
-								log.Println("InstanceController: Could not find any plan to use for update")
-							} else {
-								planName = "deploy"
+
+					if paramFound {
+						if !planFound {
+							// The parameter doesn't have a trigger, try to find the corresponding default plan.
+							names := []string{"update", "deploy"}
+							for _, n := range names {
```

Comment:
Since we're refactoring this code - wdyt about extracting this loop into a separate function:
```go
findPlan := func(names []string) (string, boolean) {
for _, n := range names {
   if plan, found = fv.Spec.Plans[n]; found {
       return plan, found
   }
   return "", false
}
```
and then calling it here (and above) with:
```go
planName, planFound := findPlan([]string{"update", "deploy"})
```

Reply:
I personally think that trying to understand the code for the first time would have to do an extra context switch to see what `findPlan()` actually does and then come back.

This would be justified if we'd `findPlan()` very often, but we only use it twice, so my *personal* opinion is that at this point extracting it into a separate function would only hurt readability.

My opinion on this matter is not very strong, so if you feel stronger about this, I can totally extract it =).

Reply:
I would probably also extract it but I don't have strong opinion here :)